MANTA-4729 Provide alternative log rotation functions for manta services

kellymclaughlin opened the pull request:
This is part of the work described by [RFD 142](https://github.com/joyent/rfd/blob/master/rfd/0142/README.md). It provides a path for manta services that run multiple instances per zone to use SMF logging rather than syslog logging and ensure those logs are properly rotated and uploaded.

There is a new `logupload.sh` script that is used to manage the entire log rotation and upload process. This is one of the goals set out in the RFD. It also introduces some randomness to the timing of the log file upload process. We currently have the problem that all manta services upload log files immediately after rotating the  files at the top of each hour. This had been observed to impact the manta service and the hope is the random wait time used by the `logupload.sh` script will stagger the uploads and diminish the impact.

### Testing

I have been testing this indirectly via the changes to incorporate this work into [`manta-buckets-api`](https://github.com/joyent/manta-buckets-api/tree/MANTA-1936). I have verified that the `logadm` configuration is correct as well as the `crontab` entries. I have also observed the log rotation and upload process and verified it is working as expected for both the `bucket-api` log files and the other service log files in the zone such as `config-agent` and `registrar`.

kellymclaughlin committed:
MANTA-1936 Rip out syslog logging

Add some new helper functions and a new logupload.sh script to manage the log
rotation and upload for services moving away from syslog logging.

kellymclaughlin committed:
Fix a couple of issues uncovered while testing changes

* Copy logupload.sh script so it can be found by logadm
* Define logadm rule pattern in manta_buckets_add_logadm_entry function

kellymclaughlin committed:
Trying to get logadm template to cooperate

kellymclaughlin committed:
Correct some more bugs found while testing

* Remove unnecessary log function from logupload.sh
* Use seconds instead of milliseconds multiplier for sleep time in logupload.sh
* Move the port number to the end of the filename for the logadm template in
manta_buckets_add_logadm_entry
* Allow logupload.sh script to work with services that only have a single
instance such as registrar.

kellymclaughlin committed:
Remove remaining use of log function in logupload.sh

kellymclaughlin committed:
Minor fix for extracting date from log files

Reinstate the sed substitution for logtime for services with only a single
instance.

bowrocker reviewed (dismissed):
LGTM.

trentm commented on services.sh:
```diff
@@ -214,6 +214,91 @@ function manta_setup_common_log_rotation {
+    manta_add_logadm_entry "config-agent"
+    manta_add_logadm_entry "registrar"
+    if [[ $# -ge 1 ]]; then
+        for port in `svcs -H -o fmri $1 | grep -v default | cut -f4 -d-`
```
A couple points here:

- Would we not eventually want this `manta_buckets_setup_common_log_rotation` to be used for "buckets-*" services other than "buckets-api" that **don't have multiple per-port SMF service instances** (i.e. to benefit of the new "logupload.sh")? If so, then this grepping out `default` won't work.  Actually is the "default" instance a bug in that the SMF manifest has `<single_instance/>` in it?  There are other parts below that also assume there is a port in play. I wonder if that can relatively easily be made not an assumption.

- This `cut -f4 -d-` is assuming the "SERVICE" arg will have exactly one hyphen in it, that there are no other hyphens in the FMRI, and that the SMF service instance name will be "$service-$port".  The latter is probably fine, the others, I think could be surprising. How about (a) documenting the "$service-$port" SMF instance assumption, (b) checking inside the for-loop block that "$port" isn't the empty string and (c) changing to `| awk -F- '{print $NF}'` instead of `cut ...`.

kellymclaughlin replied:
Yep, I gotta rework this some. Hopefully my intent is clear. I just want to get away from having to make assumption about the count of instances when there are more than one so we're not stuck with some limit strictly based on our logging configuration. The `default` log files presumably would still need to be rotated for multi-instance services so I need to handle that. Maybe I'll iterate through the instance entries and then add a line for the `default` case that should handle the single instance service case.

kellymclaughlin replied:
Ok, I think I've addressed the concerns here now. What should happen is that `logadm` rules should be created for any instances for multiple instance services. Then we also create a rule for the `default` service. This should allow the same `manta_buckets_setup_common_log_rotation` function to be used for any service.

trentm commented on services.sh:
```diff
@@ -214,6 +214,91 @@ function manta_setup_common_log_rotation {
+# manta_buckets_setup_common_log_rotation [SERVICE]: set up cron to rotate logs,
+# and then configure log rotation for log files common to all Manta zones.  If
+# SERVICE is not specified, then only the truly common logs will be set up.  If
+# SERVICE is given, then logs in /var/svc/log for each instance of SERVCIE will
```
nit: tyop "SERVCIE"

kellymclaughlin replied:
Fixed

trentm commented on services.sh:
```diff
@@ -214,6 +214,91 @@ function manta_setup_common_log_rotation {
+    # /var/log/manta/upload) and then upload after that.
+    #
+    crontab -l > /tmp/.manta_logadm_cron
+    echo '0 * * * * /opt/smartdc/common/sbin/logupload.sh >> /var/log/logupload.log 2>&1' \
```
Theoretically could have a separate "manta_common_setup_end" that didn't bother adding:

```
    logadm -w mbackup -C 3 -c -s 1m '/var/log/mbackup.log'
```

These manta-scripts bash libraries are getting a bit long in the tooth.

kellymclaughlin replied:
Yeah, good point on not having the mbackup entry when we don't need it. I can add a `manta_buckets_common_setup_end` or something similar that can be used.

kellymclaughlin replied:
Done

trentm commented on util.sh:
```diff
@@ -63,6 +63,31 @@ function manta_add_logadm_entry {
+        logdir="$3"
+    fi
+
+    local pattern="$logdir/*$service-$port*.log"
```
I think that pattern could be tightened up FWIW to:
```
local pattern="$logdir/*:$service-$port.log"
```

trentm commented on logupload.sh:
```diff
@@ -0,0 +1,211 @@
+#
+
+echo ""   # blank line in log file helps scroll btwn instances
+set -o errexit
```
No `set -o pipefail` as well?

trentm commented on logupload.sh:
```diff
@@ -0,0 +1,211 @@
+    manta_put "/logs/$1/$year" "$DIR_TYPE"
+    manta_put "/logs/$1/$year/$month" "$DIR_TYPE"
+    manta_put "/logs/$1/$year/$month/$day" "$DIR_TYPE"
+    manta_put "/logs/$1/$year/$month/$day/$hour" "$DIR_TYPE"
```
For another day, I wonder if we could reduce this load (perhaps premature optimization), by checking to see if the day dir exists already. If so, then don't do mmkdir's on the parent dirs.

kellymclaughlin replied:
Seems like a good idea, but I agree it's probably not necessary for this change.

trentm commented on logupload.sh:
```diff
@@ -0,0 +1,211 @@
+# Immutables
+
+SSH_KEY=/root/.ssh/id_rsa
+BACKUP_LOCKFILE=/tmp/backup_lockfile
```
Intentionally using the lockfile as "backup.sh"? If so, should doc that.

trentm commented on logupload.sh:
```diff
@@ -0,0 +1,211 @@
+#
+
+#
+# Push /var/log/manta/upload/... log files up to Manta.
```
I think it would be clearer to name this "logrotateandupload.sh" (or similar). I was initially surprised that it was handling the `logadm` call itself.

This top comment should be updated.

kellymclaughlin committed:
Make some updates based on review and testing

* Rework the logadm rules for multi-instance service
* Fix handling of single instance services when creating logadm rules
* Rename logupload.sh to logrotateandupload.sh

kellymclaughlin committed:
Minor fix and comment update to logrotateandupload.sh

trentm commented on logrotateandupload.sh:
```diff
@@ -0,0 +1,213 @@
+
+
+# Files look like this:
+#     buckets-api_0db94777-555d-4f1a-a87f-b1e2ee13c025_2012-10-17T21:00:00_8081.log
```
```suggestion
#     ${SERVICE}_${NODENAME}_${TIMESTAMP}[_${INSTANCE}].log
#     buckets-api_0db94777-555d-4f1a-a87f-b1e2ee13c025_2012-10-17T21:00:00_8081.log
```

trentm commented on logrotateandupload.sh:
```diff
@@ -0,0 +1,213 @@
+# or this:
+#     boray_b3a7f519-1096-4e47-9a56-efbd1ab8b692_2012-10-17T21:00:00.log
+# And we transform them to this in manta:
+#     /poseidon/stor/logs/buckets-api/2012/10/17/20/0db94777.8081.log
```
```suggestion
#     /poseidon/stor/logs/${SERVICE}/${HOURDIR}/${SHORTNODENAME}[.${INSTANCE}].log
#     /poseidon/stor/logs/buckets-api/2012/10/17/20/0db94777.8081.log
```

trentm commented on logrotateandupload.sh:
```diff
@@ -0,0 +1,213 @@
+# minutes and then proceed with uploading the log files.
+sleeptime=$((RANDOM % 11))
+echo "delaying log upload for $sleeptime minutes"
+sleep $(($sleeptime * 60))
```
Perhaps change these to:
```
sleepsecs=$((RANDOM % 600))
```

To get a random distribution over all seconds in that 10 minutes? The way it is now there will be spikes at each of the discrete 10 minute marks after the hour. With eventual hundreds of components it might be preferable to spread them more?

trentm commented on logrotateandupload.sh:
```diff
@@ -0,0 +1,213 @@
+# Immutables
+
+SSH_KEY=/root/.ssh/id_rsa
+LOCKFILE=/tmp/log_upload_lockfile
```
You changed this var name to `LOCKFILE` which means it is sharing a global var with the var name used in `function create_lockfile`. That function also *changes* that `LOCKFILE` global var value, which is gross. Here is a suggested (but untested, sorry) patch:

```diff
diff --git a/logrotateandupload.sh b/logrotateandupload.sh
index 454bc4d..195eee9 100755
--- a/logrotateandupload.sh
+++ b/logrotateandupload.sh
@@ -30,7 +30,7 @@ export PATH=/opt/local/bin:$PATH
 # Immutables

 SSH_KEY=/root/.ssh/id_rsa
-LOCKFILE=/tmp/log_upload_lockfile
+LOCKFILE_PATH=/tmp/logrotateandupload.lock

 MANTA_KEY_ID=$(ssh-keygen -l -f $SSH_KEY.pub | awk '{print $2}')
 MANTA_URL=$(json -f /opt/smartdc/common/etc/config.json manta.url)
@@ -71,11 +71,15 @@ function sign() {
 # $1 -> lockfile to create
 # $$ pid of this script
 function create_lockfile() {
+    local TEMPFILE
+    local LOCKFILE
+    local STALE_PID
+
     # Creating the tempfile can race, but
     # ln(1) is atomic, so that's the true locking
     # operation
     TEMPFILE="$1.$$"
-    LOCKFILE="$1.lock"
+    LOCKFILE="$1"
     if ! echo $$ > $TEMPFILE 2>/dev/null; then
         echo "Unable to write to directory: $(dirname $TEMPFILE)" >&2
         return 1
@@ -210,4 +214,4 @@ echo "log file upload complete"

 # Remove lockfile only if everything succeeded, otherwise it will get cleaned
 # up as a stale pid on a following run
-/usr/bin/rm -f "$LOCKFILE"
+/usr/bin/rm -f "$LOCKFILE_PATH"
```

This isn't perfect because `function finish` still assumes internal knowledge of the TEMPFILE value.

kellymclaughlin replied:
Speaking of the `finish` function it seems like due to [this](https://github.com/joyent/manta-scripts/blob/MANTA-1936/logrotateandupload.sh#L158) line that none of the stale PID stuff is actually going to work as expected and it seems like that line makes [this](https://github.com/joyent/manta-scripts/blob/MANTA-1936/logrotateandupload.sh#L213-L215) redundant.

trentm replied:
Er, I might misunderstand what the "EXIT" `trap` does, but my read of that is:

- if we exit, remove the TEMPFILE in case we crashed in the middle of `function create_lockfile`

I'm not entirely sure without playing, however. I haven't combed carefully through `create_lockfile`.

kellymclaughlin replied:
I caught my mistake. I realized this was for TEMPFILE and not the regular LOCKFILE so it makes sense to me now.

trentm commented on services.sh:
```diff
@@ -214,6 +214,99 @@ function manta_setup_common_log_rotation {
+    local DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+    mkdir -p /opt/smartdc/common/sbin
+    cp ${DIR}/backup.sh /opt/smartdc/common/sbin
+    chmod 755 /opt/smartdc/common/sbin/backup.sh
```
Perhaps drop these two lines.

trentm commented on services.sh:
```diff
@@ -214,6 +214,99 @@ function manta_setup_common_log_rotation {
+    manta_add_logadm_entry "config-agent"
+    manta_add_logadm_entry "registrar"
+    if [[ $# -ge 1 ]]; then
+        for port in `svcs -H -o fmri $1 | grep -v default | sed s/'.*-'//`
```
```suggestion
        for port in `svcs -H -o fmri $1 | grep -v :default | sed s/'.*-'//`
```

for the freak case of there possibly being the string "default" elsewhere in the FMRI.

trentm commented on services.sh:
```diff
@@ -214,6 +214,99 @@ function manta_setup_common_log_rotation {
 }
 
+#
+# manta_buckets_setup_common_log_rotation [SERVICE]: set up cron to rotate logs,
```
These new functions aren't so much specific to *buckets* as an alternative set of `manta common setup` functions that any Manta services could use, right?  I played with a renaming of these functions to `manta_common2_...` instead of `manta_buckets_...`: https://gist.github.com/trentm/ee5408279f761380c8f0cb8b87f417dd

Thoughts?

kellymclaughlin replied:
This occurred to me as well. The only alternative name proposal I can come up with is using the `_v2` suffix, but then we risk confusion with `manta v2` which could include use of both old and new functions so `common2` is good enough. Hopefully these new functions can be replaced with an actual program that will let us do better scoping and naming.

trentm replied:
yah, I avoided "mantav2" for the same reasoning.

trentm commented on logrotateandupload.sh:
```diff
@@ -0,0 +1,213 @@
+#!/bin/bash
```
"MANTA-1936 Rip out syslog logging" by itself will be a really misleading commit message for this change to manta-scripts. The lede on this should be about offering an alternative log rotation and upload scheme for manta zones to use.

kellymclaughlin replied:
I'm not quite clear what you're suggesting. Do you think we should not follow the convention of having the first line of the commit reflect the Jira issue text or that the commit description should include more details about what is going on? It is only tangentially related to MANTA-1936 so I could create a new Jira issue that described adding a functions to support an alternative log rotation approach.

trentm replied:
One of:

- a separate ticket with a more appropriate name for this change, if not too much trouble
- make sure to add prose to the commit message to give some details on the important changes to *this* repo, or
- make an exception to the "thou shalt have the first line be the ticket id and title" rule (https://www.youtube.com/watch?v=6GMkuPiIZ2k)

kellymclaughlin replied:
I made the new Jira issues and renamed this PR. I haven't changed the branch or any of the commit messages, but I will make sure the final commit message reflects the new issue.

trentm commented on services.sh:
```diff
@@ -214,6 +214,99 @@ function manta_setup_common_log_rotation {
+        do
+            pattern="$logdir/*:$service-$port.log"
+            logadm -w "$1-$port" -C 48 -c -p 1h \
+                   -t "/var/log/manta/upload/$1_\$nodename_%FT%H:00:00_$port.log" \
```
And would you also like to be a little bit of a hero and change to *not* having colons in the rotated log file names? It would take this patch:

```diff
--- a/logrotateandupload.sh
+++ b/logrotateandupload.sh
@@ -175,9 +175,9 @@ sleep $(($sleeptime * 60))


 # Files look like this:
-#     buckets-api_0db94777-555d-4f1a-a87f-b1e2ee13c025_2012-10-17T21:00:00_8081.log
+#     buckets-api_0db94777-555d-4f1a-a87f-b1e2ee13c025_20121017T210000_8081.log
 # or this:
-#     boray_b3a7f519-1096-4e47-9a56-efbd1ab8b692_2012-10-17T21:00:00.log
+#     boray_b3a7f519-1096-4e47-9a56-efbd1ab8b692_20121017T210000.log
 # And we transform them to this in manta:
 #     /poseidon/stor/logs/buckets-api/2012/10/17/20/0db94777.8081.log
 #     /poseidon/stor/logs/boray/2012/10/17/20/b3a7f519.log
@@ -200,7 +200,8 @@ do
     if [[ ! -z "$instance" ]]; then
         instance=".$instance"
     fi
-    time=$(date -d \@$(( $(date -d $logtime "+%s") - 3600 )) "+%Y/%m/%d/%H")
+    isotime=$(echo $logtime | sed -E 's/^(....)(..)(..)T(..)(..)(..)$/\1-\2-\3T\4:\5:\6/')
+    time=$(date -d \@$(( $(date -d $isotime "+%s") - 3600 )) "+%Y/%m/%d/%H")
     key="/logs/$service/$time/$zone$instance.log"
     mkdirp $service $time
     manta_put "$key" "$LOG_TYPE" "-T $f"
diff --git a/services.sh b/services.sh
index 1e5dc37..45a746f 100644
--- a/services.sh
+++ b/services.sh
@@ -286,7 +286,7 @@ function manta_buckets_setup_common_log_rotation {
         do
             pattern="$logdir/*:$service-$port.log"
             logadm -w "$1-$port" -C 48 -c -p 1h \
-                   -t "/var/log/manta/upload/$1_\$nodename_%FT%H:00:00_$port.log" \
+                   -t "/var/log/manta/upload/$1_\$nodename_%Y%m%dT%H%M%S_$port.log" \
                    "$pattern" || fatal "unable to create logadm entry"
         done
```

I could perhaps do this as a separate change just after yours, if you like. The argument *for* doing this is perhaps a little stronger for the Triton case where hourly rotated logs actually still stick around in the zone for local debugging for a number of hours, even after being uploaded.

kellymclaughlin committed:
Update logrotateandupload.sh

Co-Authored-By: Trent Mick <trent.mick@joyent.com>

kellymclaughlin committed:
Update logrotateandupload.sh

Co-Authored-By: Trent Mick <trent.mick@joyent.com>

kellymclaughlin committed:
Update services.sh

Co-Authored-By: Trent Mick <trent.mick@joyent.com>

kellymclaughlin committed:
Make some updates based on review feedback

* Rename new functions to avoid implications they are buckets-specific
* Better distribution of sleep times
* Fix some issues with LOCKFILE handling in logrotateanduplaod.sh

kellymclaughlin commented:
I've pushed some more changes to address all the feedback except the issue subject, but I'll need to test them out in an image still. I'll hit the notify button once it's ready to go again.

kellymclaughlin committed:
Rename manta_buckets_setup_common_log_rotation

Rename manta_buckets_setup_common_log_rotation to
manta_setup_common2_log_rotation.

kellymclaughlin committed:
Fix some testing issues

kellymclaughlin committed:
Fix a few more issues found while testing

kellymclaughlin commented:
I've tested the latest set of changes in a `buckets-api` image and the logs seem are being rotated and uploaded as expected for the `buckets-api` logs as well as the `registrar` and `config-agent` logs.

trentm commented on logrotateandupload.sh:
```diff
@@ -0,0 +1,216 @@
+# minutes and then proceed with uploading the log files.
+sleepsecs=$((RANDOM % 600))
+echo "delaying log upload for $sleepsecs seconds"
+sleep $(($sleepsecs))
```
```suggestion
sleep $sleepsecs
```

nit

kellymclaughlin replied:
Changed in latest commit

trentm reviewed (dismissed):
Thanks so much for taking my patches and *doing the testing and finding my issues in there*. I have a nit in logrotateandupload.sh which I'm fine handling in a separate commit if you want to get this in and going.

A side Q: was `manta_setup_common2_log_rotation` as the name intentional rather than `manta_common2_setup_log_rotation` which would result in a common `manta_common2_` prefix on at least some of the meant-to-be-used-by-setup.sh-scripts functions?  Totally fine if so.

kellymclaughlin commented:
> A side Q: was manta_setup_common2_log_rotation as the name intentional rather than manta_common2_setup_log_rotation which would result in a common manta_common2_ prefix on at least some of the meant-to-be-used-by-setup.sh-scripts functions? Totally fine if so.

It was just an iteration on the name of existing `manta_setup_common_log_rotation` function, but I'm fine to make that change to make the naming uniform.

kellymclaughlin committed:
A couple of minor updates from review feedback
